logic/scorer/scorer.go

```diff
@@ -87,29 +103,127 @@ func scoreFifteens(ptVals []int) (scoreType, int) {
 	return st, int(numFifteens * 2)
 }
 
-func howManyAddUpTo(goal int, ptVals []int) uint {
-	if len(ptVals) == 0 {
+func howManyAddUpTo(goal int, ptVals values, start int) uint {
+	if start == cardsPerHand {
 		return 0
 	}
 
 	var many uint
-	for i, o := range ptVals {
+	for i := start; i < cardsPerHand; i++ {
+		o := ptVals[i]
 		if o > goal {
-			break
+			continue
 		} else if o == goal {
 			many++
 		} else {
 			// o is less than the goal. See what we can find with it
-			subWith := howManyAddUpTo(goal-o, ptVals[i+1:])
+			subWith := howManyAddUpTo(goal-o, ptVals, i+1)
 			many += subWith
 		}
 	}
 
 	return many
 }
 
-// Assumes input is sorted and has len 5
-func scoreRunsAndPairs(values []int) (scoreType, int) { //nolint:gocyclo
+func scorePairs(valuesToCounts valueToCount) (scoreType, int) {
+	pairPoints := 0
+	var pairType scoreType
+	for n, ct := range valuesToCounts {
+		if n == 0 {
+			continue
+		}
+		switch ct {
+		case 4:
+			return quad, 12
+		case 3:
+			pairType |= triplet
+			pairPoints += 6
+		case 2:
+			pairPoints += 2
+			if pairType&onepair > 0 {
+				return twopair, 4
+			}
+			pairType |= onepair
+		}
+	}
+	if pairType == 0 {
+		return none, 0
+	}
+	return pairType, pairPoints
+}
+
+func scoreRuns(values values, valuesToCounts valueToCount) (scoreType, int) {
+	var (
+		mult   uint8
+		runLen uint8
+	)
```

Comment:
🌵 `var mult, runLen uint8`

Reply:
also we probably don't gain a ton having these be `uint8`; maybe best to just use ints